File: hardware_control_RE/mcu_parser.py
#!/usr/bin/env python3
"""
ECOVACS RE -- full MCU frame parser (binary stream -> classified + parsed frames).

Reads the raw ttys byte stream (e.g. message_dump.bin), walks frames, validates
framing + CRC8, classifies by (cmd0, cmd1), and parses each payload via the
node classes in mcu_nodes.py.

Wire frame (verified against message_dump.bin, 10030 frames, 0 CRC failures):

    frame = [0x60] [ack] [len] [cmd0] [cmd1] [data...] [crc8] [0x0A]
      len     = data_len + 2            (counts the 2 command bytes)
      crc8    = CRC-8 over (0x60, ack, len, cmd0, cmd1, data...)
                poly 0x07, init 0x00, xorout 0x00
      payload = data[0:]               (starts right after the 2 cmd bytes)

Usage:
    python mcu_parser.py message_dump.bin            # parse whole file, print
    python mcu_parser.py message_dump.bin --summary  # per-command counts
"""
import argparse, sys

from mcu_nodes import NODES

CRC_POLY   = 0x07
CRC_INIT   = 0x00
CRC_XOROUT = 0x00

# (cmd0, cmd1) -> (node key, parse method name)
CLASSIFY = {
    (0x47, 0x44): ("IMU",        "parse_GD"),   # imu/ImuSensor
    (0x57, 0x44): ("WheelNode",  "parse_WD"),   # wheel distance
    (0x4D, 0x53): ("MotorNode",  "parse_MS"),   # motor speed report
    (0x4D, 0x42): ("MotorNode",  "parse_MB"),   # motor current
    (0x57, 0x52): ("WheelNode",  "parse_WR"),   # wheel report
    (0x47, 0x53): ("IMU",        "parse_GS"),   # imu state
    (0x43, 0x43): ("BigData",    "parse_CC"),   # battery status
    (0x5A, 0x52): ("ScreenInfo", "parse_ZR"),   # rain detect
    (0x44, 0x42): ("TemperatureAlert", "parse_DB"),  # alert/alarm
    (0x54, 0x42): ("LogSave",    "parse_TB"),   # log string
    (0x42, 0x43): ("OnOffInfo",  "parse_BC"),   # sensor/GPIO status
    (0x5A, 0x54): ("ScreenInfo", "parse_ZT"),   # screen touch/control
}

def crc8(data: bytes) -> int:
    c = CRC_INIT
    for b in data:
        c ^= b
        for _ in range(8):
            c = ((c << 1) & 0xFF) ^ (CRC_POLY if c & 0x80 else 0)
    return c ^ CRC_XOROUT

def iter_frames(stream: bytes):
    """Yield (cmd0, cmd1, ack, data, ok, reason) for every 0x60-framed message."""
    i, n = 0, len(stream)
    while i < n - 7:
        if stream[i] != 0x60:
            i += 1
            continue
        ln = stream[i+2]
        end = i + ln + 5                      # 0x60,ack,len + cmd0,cmd1 + data + crc + 0x0A
        if end >= n:
            i += 1
            continue
        if stream[end-1] != 0x0A:
            i += 1
            continue
        ack = stream[i+1]
        c0, c1 = stream[i+3], stream[i+4]
        data = stream[i+5 : i+5+(ln-2)]
        stored = stream[i+5+ln-2]
        calc = crc8(bytes([0x60, ack, ln, c0, c1]) + data)
        yield c0, c1, ack, data, calc == stored, "crc" if calc != stored else ""
        i = end

def parse_frame(c0, c1, ack, data):
    """Classify + parse one frame. Returns dict with cmd/ack/parsed, or
    {'cmd':..., 'ack':..., 'unknown':True} for unclassified commands."""
    key = (c0, c1)
    cmd = chr(c0) + chr(c1) if (0x20 <= c0 < 0x7f and 0x20 <= c1 < 0x7f) else "?"
    if key in CLASSIFY:
        node, meth = CLASSIFY[key]
        fn = getattr(NODES[node], meth)
        parsed = fn(data, ack)
        return {"cmd": cmd, "ack": ack, "data": data.hex(), "parsed": parsed}
    return {"cmd": cmd, "ack": ack, "data": data.hex(), "unknown": True}

def parse_stream(stream: bytes, validate=True):
    """Parse all frames. Returns list of dicts. If validate, raises on CRC bad."""
    out = []
    for c0, c1, ack, data, ok, reason in iter_frames(stream):
        if not ok:
            raise ValueError("bad crc frame cmd=%s%c%c" % (hex(c0), c1, c1))
        out.append(parse_frame(c0, c1, ack, data))
    return out

import time
import os
class MCULink:
    def __init__(self, path, baud):
        self.path = path
        self.baud = baud
        self.ser = None

    def open(self):
        import termios
        self.fd = os.open(self.path, os.O_RDWR | os.O_NOCTTY | os.O_NONBLOCK)
        attr = termios.tcgetattr(self.fd)
        # attr = [iflag, oflag, cflag, lflag, ispeed, ospeed, cc]

        B = {9600: termios.B9600, 57600: termios.B57600,
             115200: termios.B115200}.get(self.baud, termios.B115200)

        attr[0] = 0        # iflag: raw
        attr[1] = 0        # oflag: raw
        attr[2] = termios.CS8 | termios.CREAD | termios.CLOCAL  # cflag: 8N1, no parity
        attr[3] = 0        # lflag: raw mode
        attr[4] = B        # ispeed
        attr[5] = B        # ospeed
        attr[6][termios.VMIN] = 0
        attr[6][termios.VTIME] = 0

        termios.tcsetattr(self.fd, termios.TCSANOW, attr)
        return self

    def write(self, data: bytes):
        os.write(self.fd, data)

    def read_frame(self, timeout=1.0, poll=0.05):
        """Read one full framed (0x60..0x0A) message; returns (cmd, data) or None."""
        buf = bytearray()
        end = time.time() + timeout
        while time.time() < end:
            try:
                chunk = os.read(self.fd, 256)
                if chunk:
                    buf.extend(chunk)
            except BlockingIOError:
                pass
            while True:
                i = buf.find(b"\x60")
                if i < 0:
                    buf.clear()  # no delimiter at all left; nothing to keep
                    break
                if len(buf) - i < 3:
                    break
                ln = buf[i+2]
                total = ln + 5
                if len(buf) - i < total:
                    break
                frame = bytes(buf[i:i+total])

                # Validate BEFORE trusting this as a real frame.
                terminator_ok = frame[-1] == 0x0A
                check = crc8(frame[:-2])
                crc_ok = check == frame[-2]

                if terminator_ok and crc_ok:
                    del buf[:i+total]
                    return frame[3:5], frame[5:5+ln-2], frame

                # False positive: this 0x60 wasn't a real header.
                # Only drop the one leading byte so we don't destroy
                # a genuine frame that may start later in buf.
                print(f"desync/bad frame at offset {i} "
                      f"(terminator_ok={terminator_ok}, crc_ok={crc_ok}, "
                      f"expected_crc={check}, got={frame[-2]}) - resyncing")
                del buf[:i+1]
                # loop again to find the next 0x60 candidate
            time.sleep(poll)
        return None

def main():
    ap = argparse.ArgumentParser(description="ECOVACS RE MCU frame parser")
    ap.add_argument("file", nargs="?", default="message_dump.bin")
    args = ap.parse_args()

    link = MCULink(args.file, 115200)
    link.open()
    last = {}
    try:
        while True:
            got = link.read_frame(timeout=1.0)
            if got is not None:
                cmd, data, rawframe = got
                print(f"TYPE:{str(cmd)} ACK:{rawframe[1]} LEN:{rawframe[2]}") 
                parsed = parse_frame(cmd[0],cmd[1],0,data)
                if parsed is not None:
                    print(parsed)
                else:
                    print(f"Unknown message type {cmd}: {rawframe.hex()}")
            else:
                # no new frame; let user know about timeout
                print("Nothin...")
    finally:
        os.close(link.fd)
        
if __name__ == "__main__":
    main()

File: open_goat_firmware/src/drivers/mcu/MCU_parser.cpp
#include <iostream>
#include <vector>
#include <iomanip>
#include <cstdint>
#include <string>
#include <optional>
#include <cstdint>
#include <chrono>
#include <thread>
#include <algorithm>
#include <ulog.h>
#include <fcntl.h>
#include <termios.h>
#include <unistd.h>
#include <cerrno>

// CRC8 config 
static constexpr uint8_t CRC_POLY = 0x07;
static constexpr uint8_t CRC_INIT = 0x00;
static constexpr uint8_t CRC_XOROUT = 0x00;

static uint8_t crc_table[256];

static void build_crc_table() {
	for (int i = 0; i < 256; ++i) {
		uint8_t v = static_cast<uint8_t>(i);
		for (int b = 0; b < 8; ++b) {
			v = (v & 0x80) ? static_cast<uint8_t>((v << 1) ^ CRC_POLY) : static_cast<uint8_t>(v << 1);
		}
		crc_table[i] = v;
	}
}

static uint8_t crc8(const uint8_t *data, size_t len) {
	uint8_t crc = CRC_INIT;
	for (size_t i = 0; i < len; ++i) {
		crc = crc_table[(crc ^ data[i]) & 0xFF];
	}
	return static_cast<uint8_t>(crc ^ CRC_XOROUT);
}

class MCULink {
private:
    std::string path;
    int baud;
    int fd = -1;

    speed_t get_baud_rate_constant(int baud_rate) const {
        switch (baud_rate) {
            case 9600:   return B9600;
            case 57600:  return B57600;
            case 115200: return B115200;
            default:     return B115200;
        }
    }

public:
    MCULink(std::string path, int baud) 
        : path(std::move(path)), baud(baud) {
            	// ensure CRC table is built once
            static bool crc_ready = false;
            if (!crc_ready) {
                build_crc_table();
                crc_ready = true;
            }
            open();
        }

    ~MCULink() {
        close();
    }

    // Prevent copying to avoid duplicate file descriptor management
    MCULink(const MCULink&) = delete;
    MCULink& operator=(const MCULink&) = delete;

    // Enable move semantics
    MCULink(MCULink&& other) noexcept 
        : path(std::move(other.path)), baud(other.baud), fd(other.fd) {
        other.fd = -1;
    }

    MCULink& operator=(MCULink&& other) noexcept {
        if (this != &other) {
            close();
            path = std::move(other.path);
            baud = other.baud;
            fd = other.fd;
            other.fd = -1;
        }
        return *this;
    }

    bool open() {
        fd = ::open(path.c_str(), O_RDWR | O_NOCTTY);
        if (fd < 0) {
            ULOG_ERROR("failed to open TTY for MCULink!");
            return false;
        }

        struct termios attr{};
        if (tcgetattr(fd, &attr) != 0) {
            close();
            ULOG_ERROR("failed to get TTY attributes for MCULink!");
            return false;
        }

        speed_t speed = get_baud_rate_constant(baud);

        attr.c_iflag = 0;                                 // raw input
        attr.c_oflag = 0;                                 // raw output
        attr.c_cflag = CS8 | CREAD | CLOCAL;              // 8N1, enable receiver, ignore modem control
        attr.c_lflag = 0;                                 // raw mode (non-canonical)
        
        cfsetispeed(&attr, speed);
        cfsetospeed(&attr, speed);

        attr.c_cc[VMIN] = 1;                              // blocking read, wait for at least one byte
        attr.c_cc[VTIME] = 0;

        if (tcsetattr(fd, TCSANOW, &attr) != 0) {
            ULOG_ERROR("failed to set TTY attributes for MCULink!");
            close();
            return false;
        }

        return true;
    }

    void close() {
        if (fd >= 0) {
            ::close(fd);
            fd = -1;
        }
    }


    ssize_t write(const std::vector<uint8_t>& data) {
        if (fd < 0) return -1;
        return ::write(fd, data.data(), data.size());
    }

    void write_frame(uint8_t cmd0, uint8_t cmd1, const uint8_t* data, size_t size, uint8_t ack) {
        // write [0x60][ack][len][cmd0][cmd1][data...][crc][0x0A]
        // len is payload bytes + cmd0 + cmd1 (i.e. data size + 2)
        std::vector<uint8_t> frame;
        frame.reserve(1 + 1 + 1 + 1 + 1 + size + 1 + 1);

        frame.push_back(0x60);
        frame.push_back(ack);
        frame.push_back(static_cast<uint8_t>(size + 2u));
        frame.push_back(cmd0);
        frame.push_back(cmd1);

        if (data != nullptr && size > 0) {
            frame.insert(frame.end(), data, data + size);
        }

        uint8_t crc = crc8(frame.data(), frame.size());
        frame.push_back(crc);
        frame.push_back(0x0A);

        this->write(frame);
    }

    std::optional<std::vector<uint8_t>> read_frame(double timeout = 1.0, double poll = 0.05) {
        (void)timeout;
        (void)poll;

        if (fd < 0) {
            ULOG_ERROR("MCU LOST TTY FILE, REOPENING");
            open();
            return std::nullopt;
        }

        std::vector<uint8_t> frame;
        while (true) {
            uint8_t byte = 0;
            const ssize_t n = ::read(fd, &byte, 1);
            if (n < 0) {
                if (errno == EINTR) {
                    continue;
                }
                ULOG_ERROR("MCU read failed: %s", ::strerror(errno));
                return std::nullopt;
            }
            if (n == 0) {
                continue;
            }

            if (frame.empty()) {
                if (byte != 0x60) {
                    continue;
                }
                frame.push_back(byte);
                continue;
            }

            if (frame.size() == 1) {
                frame.push_back(byte);
                continue;
            }

            if (frame.size() == 2) {
                frame.push_back(byte);
                const uint8_t length = frame[2];
                if (length < 2) {
                    frame.clear();
                    continue;
                }
                continue;
            }

            const size_t expected_total = static_cast<size_t>(frame[2]) + 5u;
            frame.push_back(byte);

            if (frame.size() < expected_total) {
                continue;
            }

            if (frame.size() > expected_total) {
                auto next_start = std::find(frame.begin() + 1, frame.end(), static_cast<uint8_t>(0x60));
                if (next_start != frame.end()) {
                    const auto offset = std::distance(frame.begin(), next_start);
                    frame.erase(frame.begin(), frame.begin() + static_cast<std::ptrdiff_t>(offset));
                    continue;
                }
                frame.clear();
                continue;
            }

            const uint8_t got_crc = frame[frame.size() - 2];
            const uint8_t terminator = frame.back();
            const uint8_t expected_crc = crc8(frame.data(), frame.size() - 2);

            if (terminator == 0x0A && expected_crc == got_crc) {
                return frame;
            }

            auto next_start = std::find(frame.begin() + 1, frame.end(), static_cast<uint8_t>(0x60));
            if (next_start != frame.end()) {
                const auto offset = std::distance(frame.begin(), next_start);
                frame.erase(frame.begin(), frame.begin() + static_cast<std::ptrdiff_t>(offset));
                continue;
            }

            frame.clear();
        }
    }
};
